File: src/lib/classBCChecker.ts
// src/lib/classBCChecker.ts
// IS 2190:2024 Table 2 — Class B/C Fire Extinguisher Requirements
//
// For flammable liquid (Class B) and flammable gas (Class C) hazards
// with shallow depth (≤6mm).
//
// Table 2 values:
//   Low hazard:      55B rating, max 300 m²/extinguisher
//   Moderate hazard: 144B rating, max 150 m²/extinguisher
//   High hazard:     233B rating, max 100 m²/extinguisher
//
// Additional rules:
//   - Min 2 per floor (exception: <100m² → 1) per cl 7.3.4
//   - ADDITIONAL to Class A complement (cl 7.1.6) — never replaces Class A
//   - Two lower-rated extinguishers shall NOT combine to meet Table 2
//     (exception: up to 3 AFFF may combine) per cl 7.3.2

import type { HazardType, ExtinguisherRequirement } from '@/types';

const CLASS_B_TABLE: Record<HazardType, { rating: string; maxAreaM2: number }> = {
    low: { rating: '55B', maxAreaM2: 300 },
    moderate: { rating: '144B', maxAreaM2: 150 },
    high: { rating: '233B', maxAreaM2: 100 },
};

export function checkClassBC(
    floorAreas: number[],
    hazardType: HazardType
): ExtinguisherRequirement[] {
    const { rating, maxAreaM2 } = CLASS_B_TABLE[hazardType];
    return floorAreas.map((area, idx) => {
        const floorLabel = idx === 0 ? 'Ground Floor' : `Floor ${idx}`;
        const calculated = Math.ceil(area / maxAreaM2);
        const minimum = area < 100 ? 1 : 2;
        return {
            fireClass: 'B' as const,
            minimumRating: rating,
            countRequired: Math.max(calculated, minimum),
            perFloor: true,
            clauseRef: `IS 2190:2024, Table 2 (${hazardType} hazard), cl 7.3.1–7.3.4`,
            note: `Additional to Class A complement. ${floorLabel}: area ${area}m²`,
        };
    });
}

File: backend/class_bc_checker.py
# backend/class_bc_checker.py
# IS 2190:2024 Table 2 — Class B/C Fire Extinguisher Requirements
#
# Table 2 values:
#   Low hazard:      55B rating,  max 300 m²/extinguisher
#   Moderate hazard: 144B rating, max 150 m²/extinguisher
#   High hazard:     233B rating, max 100 m²/extinguisher
#
# Additional rules:
#   - Min 2 per floor (exception: <100m² → 1) per cl 7.3.4
#   - ADDITIONAL to Class A complement (cl 7.1.6)

import math
from models import HazardType, ExtinguisherRequirement

CLASS_B_TABLE: dict[str, dict] = {
    "low": {"rating": "55B", "max_area_m2": 300},
    "moderate": {"rating": "144B", "max_area_m2": 150},
    "high": {"rating": "233B", "max_area_m2": 100},
}


def check_class_bc(
    floor_areas: list[float], hazard_type: HazardType
) -> list[ExtinguisherRequirement]:
    entry = CLASS_B_TABLE[hazard_type]
    rating = entry["rating"]
    max_area = entry["max_area_m2"]

    results: list[ExtinguisherRequirement] = []
    for idx, area in enumerate(floor_areas):
        floor_label = "Ground Floor" if idx == 0 else f"Floor {idx}"
        calculated = math.ceil(area / max_area)
        minimum = 1 if area < 100 else 2

        results.append(
            ExtinguisherRequirement(
                fire_class="B",
                minimum_rating=rating,
                count_required=max(calculated, minimum),
                per_floor=True,
                clause_ref=f"IS 2190:2024, Table 2 ({hazard_type} hazard), cl 7.3.1–7.3.4",
                note=f"Additional to Class A complement. {floor_label}: area {area}m²",
            )
        )

    return results
